docs/development/design/build-images.rst

```diff
@@ -191,94 +299,122 @@ Examples of these versions are:
 
 This case will introduce a new ``base`` image. Example, ``ubuntu22-base`` in 2022.
 Note that these images will be completely isolated from the rest and don't require them to rebuild.
-This also allow us to test new Ubuntu versions without breaking people's builds.
+This also allow us to start testing a newer Ubuntu version (e.g. 22.04 LTS) without breaking people's builds,
+even before it's officially released.
 
 How do we add an extra requirement?
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 In case we need to add an extra requirement to the ``base`` image,
 we will need to rebuild all of them.
 The new image *may have different package versions* since there may be updates on the Ubuntu repositories.
-This conveys some small risk here, but in general we shouldn't require to add packages to the base images.
+This conveys some risk here, but in general we shouldn't require to add packages to the base images.
 
-Users with specific requirements could use ``build.system_packages`` and/or ``build.extras`` in the config file.
+In case we need an extra requirement for *all our images*,
+I'd recommend to add it when creating a new base image.
 
-How do we remove an old Python version?
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+If it's strongly needed and we can't wait for a new base image,
+we could install it at build time in a similar way as we do with ``build.apt_packages`` as a temporal workaround.
 
-At some point an old version of Python will be deprecated (eg. 3.4) and will be removed.
-To achieve this, we can just remove the Docker image affected: ``ubuntu20-py34``,
-once there are no users depending on it anymore.
+How do we create a mirror for each language?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A mirror can be created with ``wget`` together with ``rclone``:
+
+#. Download all the files from the official mirror:
+
+   .. code:: bash
+
+      # https://stackoverflow.com/questions/29802579/create-private-mirror-of-http-nodejs-org-dist
+      wget --mirror --convert-links --adjust-extension --page-requisites --no-parent -e robots=off http://nodejs.org/dist
+
+#. Upload all the files to S3:
+
+   .. code:: bash
+
+      # https://rclone.org/s3/
+      rclone sync -i nodejs.org s3:languages
+
+
+.. note::
+
+   Downloading a copy of the official mirror took 15m and 52Gb.
+
+How local development will work with the new approach?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Local development will require scripts to clone the official mirrors for each language and upload them to MinIO (S3).
+Besides, a script to define a set of Python version, pre-compile them and also upload them to S3.
```

Comment:
Won't local dev just use the production mirrors? I guess for offline use, but builds don't really work offline locally very well. I guess with a local git repo they will..

This definitely seems like an edge case tho.

Reply:
I'm not referring to offline use here. Using production mirrors may be fine for those languages that provide binaries. However, we definitely need the local cache for pre-compiled Python versions.

Also, we will need the scripts to clone the official mirrors and upload them to S3 (production). I'm thinking we will do exactly the same locally but using MinIO instead with a reduced version list (only 2 or 3). There is no need to download ~100Mb each time we need to test a build locally

Reply:
Makes sense.

Reply:
> Also, we will need the scripts to clone the official mirrors and upload them to S3 (production). I'm thinking we will do exactly the same locally but using MinIO instead with a reduced version list (only 2 or 3).

This is done at https://github.com/readthedocs/readthedocs.org/pull/8453 with the bash script linked in the description.